File: featomic/examples/compute-soap.cpp
#include <iostream>
#include <chemfiles.hpp>
#include <featomic.hpp>

static std::vector<featomic::SimpleSystem> read_systems(const std::string& path);

int main(int argc, char* argv[]) {
    if (argc < 2) {
        std::cout << "error: expected a command line argument" << std::endl;
        return 1;
    }
    auto systems = read_systems(argv[1]);

    // pass hyper-parameters as JSON
    const char* parameters = R"({
        "cutoff": {
            "radius": 5.0,
            "smoothing": {"type": "ShiftedCosine", "width": 0.5}
        },
        "density": {
            "type": "Gaussian",
            "width": 0.3
        },
        "basis": {
            "type": "TensorProduct",
            "max_angular": 6,
            "radial": {"type": "Gto", "max_radial": 6}
        }
    })";

    // create the calculator with its name and parameters
    auto calculator = featomic::Calculator("soap_power_spectrum", parameters);

    // setup the calculation options
    auto options = featomic::CalculationOptions();
    options.use_native_system = true;
    options.gradients.push_back("positions");

    // run the calculation
    auto descriptor = calculator.compute(systems, options);

    // The descriptor is a metatensor `TensorMap`, containing multiple blocks.
    // We can transform it to a single block containing a dense representation,
    // with one sample for each atom-centered environment.
    descriptor.keys_to_samples("center_type");
    descriptor.keys_to_properties(std::vector<std::string>{"neighbor_1_type", "neighbor_2_type"});

    // extract values from the descriptor in the only remaining block
    auto block = descriptor.block_by_id(0);
    auto values = block.values();

    // you can now use values as the input of a machine learning algorithm

    return 0;
}


std::vector<featomic::SimpleSystem> read_systems(const std::string& path) {
    auto trajectory = chemfiles::Trajectory(path);
    auto systems = std::vector<featomic::SimpleSystem>();
    for (size_t step=0; step<trajectory.nsteps(); step++) {
        auto frame = trajectory.read_step(step);

        auto matrix = frame.cell().matrix();
        auto system = featomic::SimpleSystem(featomic::System::CellMatrix{{
            {matrix[0][0], matrix[0][1], matrix[0][2]},
            {matrix[1][0], matrix[1][1], matrix[1][2]},
            {matrix[2][0], matrix[2][1], matrix[2][2]},
        }});

        auto positions = frame.positions();
        for (size_t i=0; i<frame.size(); i++) {
            system.add_atom(
                static_cast<int32_t>(frame[i].atomic_number().value_or(0)),
                {positions[i][0], positions[i][1], positions[i][2]}
            );
        }

        systems.push_back(system);
    }

    return systems;
}

File: featomic/examples/compute-soap.rs
use metatensor::Labels;
use featomic::{Calculator, System, CalculationOptions};
use chemfiles::{Trajectory, Frame};

fn read_systems_from_file(path: &str) -> Vec<Box<dyn System>> {
    let mut trajectory = Trajectory::open(path, 'r').expect("could not open the trajectory");
    let mut frame = Frame::new();
    let mut systems = Vec::new();
    for step in 0..trajectory.nsteps() {
        trajectory.read_step(step, &mut frame).expect("failed to read single frame");
        systems.push((&frame).into());
    }

    systems
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    // load the systems from command line argument
    let path = std::env::args().nth(1).expect("expected a command line argument");
    let mut systems = read_systems_from_file(&path);

    // pass hyper-parameters as JSON
    let parameters = r#"{
        "cutoff": {
            "radius": 5.0,
            "smoothing": {
                "type": "ShiftedCosine",
                "width": 0.5
            }
        },
        "density": {
            "type": "Gaussian",
            "width": 0.3
        },
        "basis": {
            "type": "TensorProduct",
            "max_angular": 4,
            "radial": {"type": "Gto", "max_radial": 6}
        }
    }"#;
    // create the calculator with its name and parameters
    let mut calculator = Calculator::new("soap_power_spectrum", parameters.to_owned())?;

    // create the options for a single calculation, here we request the
    // calculation of gradients with respect to positions
    let options = CalculationOptions {
        gradients: &["positions"],
        ..Default::default()
    };

    // run the calculation
    let descriptor = calculator.compute(&mut systems, options)?;

    // Transform the descriptor to dense representation, with one sample for
    // each atom-centered environment, and the neighbor atomic types part of the
    // properties
    let keys_to_move = Labels::empty(vec!["center_type"]);
    let descriptor = descriptor.keys_to_samples(&keys_to_move, /* sort_samples */ true)?;

    let keys_to_move = Labels::empty(vec!["neighbor_1_type", "neighbor_2_type"]);
    let descriptor = descriptor.keys_to_properties(&keys_to_move, /* sort_samples */ true)?;

    // descriptor now contains a single block, which can be used as the input
    // to standard ML algorithms
    let values = descriptor.block_by_id(0).values().to_array();
    println!("SOAP representation shape: {:?}", values.shape());

    Ok(())
}
